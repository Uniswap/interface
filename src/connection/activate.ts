import { sendAnalyticsEvent } from '@uniswap/analytics'
import { InterfaceEventName, WalletConnectionResult } from '@uniswap/analytics-events'
import { Connection } from 'connection/types'
import { atom } from 'jotai'
import { useAtomValue, useUpdateAtom } from 'jotai/utils'
import { useCallback } from 'react'
import { useLocation } from 'react-router-dom'
import { useAppDispatch } from 'state/hooks'
import { updateSelectedWallet } from 'state/user/reducer'
import { getCurrentPageFromLocation } from 'utils/urlRoutes'

import { didUserReject } from './utils'

export enum ActivationStatus {
  PENDING,
  ERROR,
  IDLE,
}

type ActivationPendingState = { status: ActivationStatus.PENDING; connection: Connection }
type ActivationErrorState = { status: ActivationStatus.ERROR; connection: Connection; error: any }
const IDLE_ACTIVATION_STATE = { status: ActivationStatus.IDLE } as const
type ActivationState = ActivationPendingState | ActivationErrorState | typeof IDLE_ACTIVATION_STATE

const activationStateAtom = atom<ActivationState>(IDLE_ACTIVATION_STATE)

function useTryActivation() {
  const dispatch = useAppDispatch()
  const setActivationState = useUpdateAtom(activationStateAtom)
  const { pathname } = useLocation()
  const currentPage = getCurrentPageFromLocation(pathname)

  return useCallback(
    async (connection: Connection, onSuccess: () => void) => {
      // Skips wallet connection if the connection should override the default
      // behavior, i.e. install MetaMask or launch Coinbase app
      if (connection.overrideActivate?.()) return

      try {
        setActivationState({ status: ActivationStatus.PENDING, connection })

        console.debug(`Connection activating: ${connection.getName()}`)
        await connection.connector.activate()

        console.debug(`Connection activated: ${connection.getName()}`)
        dispatch(updateSelectedWallet({ wallet: connection.type }))

        // Clears pending connection state
        setActivationState(IDLE_ACTIVATION_STATE)

        onSuccess()
      } catch (error) {
        // Gracefully handles errors from the user rejecting a connection attempt
        if (didUserReject(connection, error)) {
          setActivationState(IDLE_ACTIVATION_STATE)
          return
        }

        // TODO(WEB-3162): re-add special treatment for already-pending injected errors & move debug to after didUserReject() check
        console.debug(`Connection failed: ${connection.getName()}`)
        console.error(error)

        // Failed Connection events are logged here, while successful ones are logged by Web3Provider
        sendAnalyticsEvent(InterfaceEventName.WALLET_CONNECT_TXN_COMPLETED, {
          result: WalletConnectionResult.FAILED,
          wallet_type: connection.getName(),
          page: currentPage,
        })
        setActivationState({ status: ActivationStatus.ERROR, connection, error })
      }
    },
    [currentPage, dispatch, setActivationState]
  )
}

function useCancelActivation() {
  const setActivationState = useUpdateAtom(activationStateAtom)
  return useCallback(
    () =>
      setActivationState((activationState) => {
        if (activationState.status !== ActivationStatus.IDLE) activationState.connection.connector.deactivate?.()
        return IDLE_ACTIVATION_STATE
      }),
    [setActivationState]
  )
}

export function useActivationState() {
  const activationState = useAtomValue(activationStateAtom)
  const tryActivation = useTryActivation()
  const cancelActivation = useCancelActivation()

  return { activationState, tryActivation, cancelActivation }
}
