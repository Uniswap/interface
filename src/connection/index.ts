import { CoinbaseWallet } from '@web3-react/coinbase-wallet'
import { initializeConnector, Web3ReactHooks } from '@web3-react/core'
import { GnosisSafe } from '@web3-react/gnosis-safe'
import { MetaMask } from '@web3-react/metamask'
import { Network } from '@web3-react/network'
import { Connector } from '@web3-react/types'
import COINBASE_ICON_URL from 'assets/images/coinbaseWalletIcon.svg'
import GNOSIS_ICON_URL from 'assets/images/gnosis.png'
import METAMASK_ICON_URL from 'assets/images/metamask.svg'
import UNIWALLET_ICON_URL from 'assets/images/uniwallet.svg'
import WALLET_CONNECT_ICON_URL from 'assets/images/walletConnectIcon.svg'
import INJECTED_DARK_ICON_URL from 'assets/svg/browser-wallet-dark.svg'
import INJECTED_LIGHT_ICON_URL from 'assets/svg/browser-wallet-light.svg'
import UNISWAP_LOGO_URL from 'assets/svg/logo.svg'
import { SupportedChainId } from 'constants/chains'
import { useCallback } from 'react'
import { useIsDarkMode } from 'theme/components/ThemeToggle'
import { isMobile, isNonIOSPhone } from 'utils/userAgent'

import { RPC_URLS } from '../constants/networks'
import { RPC_PROVIDERS } from '../constants/providers'
import { isCoinbaseWallet, isInjected, isMetaMaskWallet } from './utils'
import { UniwalletConnect, WalletConnectPopup } from './WalletConnect'

export enum ConnectionType {
  UNIWALLET = 'UNIWALLET',
  INJECTED = 'INJECTED',
  COINBASE_WALLET = 'COINBASE_WALLET',
  WALLET_CONNECT = 'WALLET_CONNECT',
  NETWORK = 'NETWORK',
  GNOSIS_SAFE = 'GNOSIS_SAFE',
}

export interface Connection {
  name: string
  connector: Connector
  hooks: Web3ReactHooks
  type: ConnectionType
  icon?: string
  shouldDisplay?: boolean
  overrideActivate?: () => void
  isNew?: boolean
}

function onError(error: Error) {
  console.debug(`web3-react error: ${error}`)
}

const [web3Network, web3NetworkHooks] = initializeConnector<Network>(
  (actions) => new Network({ actions, urlMap: RPC_PROVIDERS, defaultChainId: 1 })
)
export const networkConnection: Connection = {
  name: 'Network',
  connector: web3Network,
  hooks: web3NetworkHooks,
  type: ConnectionType.NETWORK,
  shouldDisplay: false,
}

const isCoinbaseWalletBrowser = isMobile && isCoinbaseWallet
const isMetaMaskBrowser = isMobile && isMetaMaskWallet
const getIsInjectedMobileBrowser = isCoinbaseWalletBrowser || isMetaMaskBrowser

const getShouldAdvertiseMetaMask = !isMetaMaskWallet && !isMobile && (!isInjected || isCoinbaseWallet)
const isGenericInjector = isInjected && !isMetaMaskWallet && !isCoinbaseWallet

const [web3Injected, web3InjectedHooks] = initializeConnector<MetaMask>((actions) => new MetaMask({ actions, onError }))
const baseInjectedConnection: Omit<Connection, 'icon'> = {
  name: isGenericInjector ? 'Browser Wallet' : 'MetaMask',
  connector: web3Injected,
  hooks: web3InjectedHooks,
  type: ConnectionType.INJECTED,
  shouldDisplay: isMetaMaskWallet || getShouldAdvertiseMetaMask || isGenericInjector,
  // If on non-injected, non-mobile browser, prompt user to install Metamask
  overrideActivate: getShouldAdvertiseMetaMask ? () => window.open('https://metamask.io/', 'inst_metamask') : undefined,
}

export const darkInjectedConnection: Connection = {
  ...baseInjectedConnection,
  icon: isGenericInjector ? INJECTED_DARK_ICON_URL : METAMASK_ICON_URL,
}

export const lightInjectedConnection: Connection = {
  ...baseInjectedConnection,
  icon: isGenericInjector ? INJECTED_LIGHT_ICON_URL : METAMASK_ICON_URL,
}

const [web3GnosisSafe, web3GnosisSafeHooks] = initializeConnector<GnosisSafe>((actions) => new GnosisSafe({ actions }))
export const gnosisSafeConnection: Connection = {
  name: 'Gnosis Safe',
  connector: web3GnosisSafe,
  hooks: web3GnosisSafeHooks,
  type: ConnectionType.GNOSIS_SAFE,
  icon: GNOSIS_ICON_URL,
  shouldDisplay: false,
}

const [web3WalletConnect, web3WalletConnectHooks] = initializeConnector<WalletConnectPopup>(
  (actions) => new WalletConnectPopup({ actions, onError })
)
export const walletConnectConnection: Connection = {
  name: 'WalletConnect',
  connector: web3WalletConnect,
  hooks: web3WalletConnectHooks,
  type: ConnectionType.WALLET_CONNECT,
  icon: WALLET_CONNECT_ICON_URL,
  shouldDisplay: !getIsInjectedMobileBrowser,
}

const [web3UniwalletConnect, web3UniwalletConnectHooks] = initializeConnector<UniwalletConnect>(
  (actions) => new UniwalletConnect({ actions, onError })
)
export const uniwalletConnectConnection: Connection = {
  name: 'Uniswap Wallet',
  connector: web3UniwalletConnect,
  hooks: web3UniwalletConnectHooks,
  type: ConnectionType.UNIWALLET,
  icon: UNIWALLET_ICON_URL,
  shouldDisplay: Boolean(!getIsInjectedMobileBrowser && !isNonIOSPhone),
  isNew: true,
}

const [web3CoinbaseWallet, web3CoinbaseWalletHooks] = initializeConnector<CoinbaseWallet>(
  (actions) =>
    new CoinbaseWallet({
      actions,
      options: {
        url: RPC_URLS[SupportedChainId.MAINNET][0],
        appName: 'Uniswap',
        appLogoUrl: UNISWAP_LOGO_URL,
        reloadOnDisconnect: false,
      },
      onError,
    })
)

export const coinbaseWalletConnection: Connection = {
  name: 'Coinbase Wallet',
  connector: web3CoinbaseWallet,
  hooks: web3CoinbaseWalletHooks,
  type: ConnectionType.COINBASE_WALLET,
  icon: COINBASE_ICON_URL,
  shouldDisplay: Boolean((isMobile && !getIsInjectedMobileBrowser) || !isMobile || isCoinbaseWalletBrowser),
  // If on a mobile browser that isn't the coinbase wallet browser, deeplink to the coinbase wallet app
  overrideActivate:
    isMobile && !getIsInjectedMobileBrowser
      ? () => window.open('https://go.cb-w.com/mtUDhEZPy1', 'cbwallet')
      : undefined,
}

export function getConnections(isDarkMode: boolean) {
  return [
    uniwalletConnectConnection,
    isDarkMode ? darkInjectedConnection : lightInjectedConnection,
    walletConnectConnection,
    coinbaseWalletConnection,
    gnosisSafeConnection,
    networkConnection,
  ]
}

export function useConnections() {
  const isDarkMode = useIsDarkMode()
  return getConnections(isDarkMode)
}

export function useGetConnection() {
  const isDarkMode = useIsDarkMode()
  return useCallback(
    (c: Connector | ConnectionType) => {
      if (c instanceof Connector) {
        const connection = getConnections(isDarkMode).find((connection) => connection.connector === c)
        if (!connection) {
          throw Error('unsupported connector')
        }
        return connection
      } else {
        switch (c) {
          case ConnectionType.INJECTED:
            return isDarkMode ? darkInjectedConnection : lightInjectedConnection
          case ConnectionType.COINBASE_WALLET:
            return coinbaseWalletConnection
          case ConnectionType.WALLET_CONNECT:
            return walletConnectConnection
          case ConnectionType.UNIWALLET:
            return uniwalletConnectConnection
          case ConnectionType.NETWORK:
            return networkConnection
          case ConnectionType.GNOSIS_SAFE:
            return gnosisSafeConnection
        }
      }
    },
    [isDarkMode]
  )
}
