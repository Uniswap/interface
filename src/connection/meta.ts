import { ConnectionType, toConnectionType } from './types'

export interface ConnectionMeta {
  type: ConnectionType
  address?: string
  ENSName?: string
}

export const connectionMetaKey = 'connection_meta'

export function getPersistedConnectionMeta(): ConnectionMeta | undefined {
  try {
    const value = localStorage.getItem(connectionMetaKey)
    if (value) {
      const raw = JSON.parse(value) as ConnectionMeta
      const connectionType = toConnectionType(raw.type)
      if (connectionType) {
        return {
          type: connectionType,
          address: raw.address,
          ENSName: raw.ENSName,
        }
      }
    }
  } catch (e) {
    console.warn(e)
  }
  return
}

export function setPersistedConnectionMeta(meta: ConnectionMeta) {
  localStorage.setItem(connectionMetaKey, JSON.stringify(meta))
}

export function deletePersistedConnectionMeta() {
  localStorage.removeItem(connectionMetaKey)
}
