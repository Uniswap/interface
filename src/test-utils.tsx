import { render, RenderOptions } from '@testing-library/react'
import React, { FC, ReactElement, ReactNode } from 'react'
import { Provider } from 'react-redux'
import store from 'state'
import ThemeProvider from 'theme'

const WithProviders: FC = ({ children }: { children?: ReactNode }) => {
  return (
    <Provider store={store}>
      <ThemeProvider>{children}</ThemeProvider>
    </Provider>
  )
}

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: WithProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
