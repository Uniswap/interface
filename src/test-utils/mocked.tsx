/**
 * Casts the passed function as a jest.Mock.
 * Use this in combination with jest.mock() to safely access functions from mocked modules.
 *
 * @example
 *
 *  import { useExample } from 'example'
 *  jest.mock('example', () => ({ useExample: jest.fn() }))
 *  beforeEach(() => {
 *    asMock(useExample).mockImplementation(() => ...)
 *  })
 */
// jest expects mocks to be coerced (eg fn as jest.MockedFunction<T>), but this is not ergonomic when using ASI.
// Instead, we use this utility function to improve readability and add a check to ensure the function is a mock.
export function mocked<T extends (...args: any) => any>(fn: T) {
  if (!jest.isMockFunction(fn)) throw new Error('fn is not a mock')
  return fn as jest.MockedFunction<T>
}
