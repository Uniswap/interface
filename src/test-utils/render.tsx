import { MockedProvider } from '@apollo/client/testing'
import { i18n } from '@lingui/core'
import { I18nProvider } from '@lingui/react'
import { queries } from '@testing-library/dom'
import { render, renderHook, RenderHookOptions, RenderOptions } from '@testing-library/react'
import { DEFAULT_LOCALE } from 'constants/locales'
import { BlockNumberProvider } from 'lib/hooks/useBlockNumber'
import catalog from 'locales/en-US'
import { en } from 'make-plural/plurals'
import { ReactElement, ReactNode } from 'react'
import { QueryClient, QueryClientProvider } from 'react-query'
import { Provider } from 'react-redux'
import { HashRouter } from 'react-router-dom'
import store from 'state'
import ThemeProvider from 'theme'

i18n.load({
  [DEFAULT_LOCALE]: catalog.messages,
})
i18n.loadLocaleData({
  [DEFAULT_LOCALE]: { plurals: en },
})
i18n.activate(DEFAULT_LOCALE)
const MockedI18nProvider = ({ children }: any) => <I18nProvider i18n={i18n}>{children}</I18nProvider>

const queryClient = new QueryClient()

const WithProviders = ({ children }: { children?: ReactNode }) => {
  return (
    <MockedI18nProvider>
      <Provider store={store}>
        <QueryClientProvider client={queryClient}>
          <HashRouter>
            {/*
             * Web3Provider is mocked through setupTests.ts
             * To test behavior that depends on Web3Provider, use jest.unmock('@web3-react/core')
             */}
            <MockedProvider>
              <BlockNumberProvider>
                <ThemeProvider>{children}</ThemeProvider>
              </BlockNumberProvider>
            </MockedProvider>
          </HashRouter>
        </QueryClientProvider>
      </Provider>
    </MockedI18nProvider>
  )
}

type CustomRenderOptions = Omit<RenderOptions, 'wrapper'>
const customRender = (ui: ReactElement, options?: CustomRenderOptions) => {
  return render<typeof queries>(ui, { ...options, wrapper: WithProviders })
}

type CustomRenderHookOptions<Props> = Omit<RenderHookOptions<Props>, 'wrapper'>
const customRenderHook = <Result, Props>(
  hook: (initialProps: Props) => Result,
  options?: CustomRenderHookOptions<Props>
) => {
  return renderHook(hook, { ...options, wrapper: WithProviders })
}

export * from '@testing-library/react'
export { customRender as render }
export { customRenderHook as renderHook }
