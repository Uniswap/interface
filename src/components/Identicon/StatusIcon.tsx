import { useWeb3React } from '@web3-react/core'
import { MouseoverTooltip } from 'components/Tooltip'
import { Unicon } from 'components/Unicon'
import { Connection, ConnectionType } from 'connection'
import useENSAvatar from 'hooks/useENSAvatar'
import { useIsMobile } from 'nft/hooks'
import { PropsWithChildren } from 'react'
import styled from 'styled-components/macro'
import { ThemedText } from 'theme'
import { flexColumnNoWrap } from 'theme/styles'

import sockImg from '../../assets/svg/socks.svg'
import { useHasSocks } from '../../hooks/useSocksBalance'
import Identicon from '../Identicon'

export const IconWrapper = styled.div<{ size?: number }>`
  position: relative;
  ${flexColumnNoWrap};
  align-items: center;
  justify-content: center;
  margin-right: 8px;
  & > img,
  span {
    height: ${({ size }) => (size ? size + 'px' : '32px')};
    width: ${({ size }) => (size ? size + 'px' : '32px')};
  }
  ${({ theme }) => theme.deprecated_mediaWidth.deprecated_upToMedium`
    align-items: flex-end;
  `};
`

const MiniIconContainer = styled.div<{ side: 'left' | 'right' }>`
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 16px;
  height: 16px;
  bottom: -4px;
  ${({ side }) => `${side === 'left' ? 'left' : 'right'}: -4px;`}
  border-radius: 50%;
  outline: 2px solid ${({ theme }) => theme.backgroundSurface};
  outline-offset: -0.1px;
  background-color: ${({ theme }) => theme.backgroundSurface};
  overflow: hidden;
  @supports (overflow: clip) {
    overflow: clip;
  }
`

const MiniImg = styled.img`
  width: 16px;
  height: 16px;
`

const Socks = () => {
  return (
    <MiniIconContainer side="left">
      <MiniImg src={sockImg} />
    </MiniIconContainer>
  )
}

const MiniWalletIcon = ({ connection, side }: { connection: Connection; side: 'left' | 'right' }) => {
  return (
    <MiniIconContainer side={side}>
      <MiniImg src={connection.icon} alt={`${connection.name} icon`} />
    </MiniIconContainer>
  )
}

const Divider = styled.div`
  border-bottom: 1px solid ${({ theme }) => theme.backgroundOutline};
  margin: 12px 0;
`

function UniconTooltip({ children, enabled }: PropsWithChildren<{ enabled?: boolean }>) {
  return (
    <MouseoverTooltip
      offsetY={8}
      disableHover={!enabled}
      text={
        // TODO(cartcrom): add Learn More link when unicon microsite is polished
        <>
          <ThemedText.SubHeaderSmall color="textPrimary" paddingTop="4px">
            This is your Unicon
          </ThemedText.SubHeaderSmall>
          <Divider />
          <ThemedText.Caption paddingBottom="4px">
            Unicons are avatars for your wallet, generated from your address.
          </ThemedText.Caption>
        </>
      }
      placement="bottom"
    >
      <div>{children}</div>
    </MouseoverTooltip>
  )
}

const MainWalletIcon = ({
  connection,
  size,
  enableInfotips,
}: {
  connection: Connection
  size: number
  enableInfotips?: boolean
}) => {
  const { account } = useWeb3React()
  const { avatar } = useENSAvatar(account ?? undefined)
  const isMobile = useIsMobile()

  if (!account) {
    return null
  } else if (avatar || (connection.type === ConnectionType.INJECTED && connection.name === 'MetaMask')) {
    return <Identicon size={size} />
  } else {
    return isMobile ? (
      <Unicon address={account} size={size} />
    ) : (
      <UniconTooltip enabled={enableInfotips}>
        <Unicon address={account} size={size} />
      </UniconTooltip>
    )
  }
}

export default function StatusIcon({
  connection,
  size = 16,
  enableInfotips,
  showMiniIcons = true,
}: {
  connection: Connection
  size?: number
  enableInfotips?: boolean
  showMiniIcons?: boolean
}) {
  const hasSocks = useHasSocks()

  return (
    <IconWrapper size={size}>
      {hasSocks && showMiniIcons && <Socks />}
      <MainWalletIcon connection={connection} size={size} enableInfotips={enableInfotips} />
      {showMiniIcons && <MiniWalletIcon connection={connection} side="right" />}
    </IconWrapper>
  )
}
