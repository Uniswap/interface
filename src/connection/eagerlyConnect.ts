import { Connector } from '@web3-react/types'
import { useSyncExternalStore } from 'react'

import { deprecatedNetworkConnection, getConnection, gnosisSafeConnection } from './index'
import { deletePersistedConnectionMeta, getPersistedConnectionMeta } from './meta'
import { ConnectionType } from './types'

class FailedToConnect extends Error {}

let connectionReady: Promise<void> | true = true

export function useConnectionReady() {
  return useSyncExternalStore(
    (onStoreChange) => {
      if (connectionReady instanceof Promise) {
        connectionReady.finally(onStoreChange)
      }
      return () => undefined
    },
    () => connectionReady === true
  )
}

async function connect(connector: Connector, type: ConnectionType) {
  performance.mark(`web3:connect:${type}:start`)
  try {
    if (connector.connectEagerly) {
      await connector.connectEagerly()
    } else {
      await connector.activate()
    }
    return true
  } catch (error) {
    console.debug(`web3-react eager connection error: ${error}`)
    return false
  } finally {
    performance.measure(`web3:connect:${type}`, `web3:connect:${type}:start`)
  }
}

// The Safe connector will only work from safe.global, which iframes;
// it is only necessary to try (and to load all the deps) if we are in an iframe.
if (window !== window.parent) {
  connect(gnosisSafeConnection.connector, ConnectionType.GNOSIS_SAFE)
}
connect(deprecatedNetworkConnection.connector, ConnectionType.DEPRECATED_NETWORK)

// Get the persisted wallet type from the last session.
const meta = getPersistedConnectionMeta()
if (meta?.type) {
  const selectedConnection = getConnection(meta.type)
  if (selectedConnection) {
    connectionReady = connect(selectedConnection.connector, meta.type)
      .then((connected) => {
        if (!connected) throw new FailedToConnect()
      })
      .catch((error) => {
        // Clear the persisted wallet type if it failed to connect.
        deletePersistedConnectionMeta()
        // Log it if it threw an unknown error.
        if (!(error instanceof FailedToConnect)) {
          console.error(error)
        }
      })
      .finally(() => {
        connectionReady = true
      })
  }
}
