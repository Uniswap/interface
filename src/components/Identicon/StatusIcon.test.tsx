import { useWeb3React } from '@web3-react/core'
import { getConnections } from 'connection'
import { mocked } from 'test-utils/mocked'
import { render } from 'test-utils/render'

import StatusIcon from './StatusIcon'

jest.mock('../../hooks/useSocksBalance', () => ({
  useHasSocks: () => true,
}))

describe('StatusIcon', () => {
  describe('with no account', () => {
    it('renders children in correct order', () => {
      const supportedConnections = getConnections()
      const injectedConnection = supportedConnections[1]
      const component = render(<StatusIcon connection={injectedConnection} />)
      expect(component.getByTestId('StatusIconRoot')).toMatchSnapshot()
    })

    it('renders without mini icons', () => {
      const supportedConnections = getConnections()
      const injectedConnection = supportedConnections[1]
      const component = render(<StatusIcon connection={injectedConnection} showMiniIcons={false} />)
      expect(component.getByTestId('StatusIconRoot').children.length).toEqual(0)
    })
  })

  describe('with account', () => {
    beforeEach(() => {
      mocked(useWeb3React).mockReturnValue({
        account: '0x52270d8234b864dcAC9947f510CE9275A8a116Db',
        isActive: true,
      } as ReturnType<typeof useWeb3React>)
    })

    it('renders children in correct order', () => {
      const supportedConnections = getConnections()
      const injectedConnection = supportedConnections[1]
      const component = render(<StatusIcon connection={injectedConnection} />)
      expect(component.getByTestId('StatusIconRoot')).toMatchSnapshot()
    })

    it('renders without mini icons', () => {
      const supportedConnections = getConnections()
      const injectedConnection = supportedConnections[1]
      const component = render(<StatusIcon connection={injectedConnection} showMiniIcons={false} />)
      expect(component.getByTestId('StatusIconRoot').children.length).toEqual(1)
    })
  })
})
