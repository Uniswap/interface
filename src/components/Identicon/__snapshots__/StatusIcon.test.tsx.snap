// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusIcon with account renders children in correct order 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-right: 8px;
}

.c0 > img,
.c0 span {
  height: 16px;
  width: 16px;
}

.c1 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 16px;
  height: 16px;
  bottom: -4px;
  right: -4px;
  border-radius: 50%;
  outline: 2px solid #FFFFFF;
  outline-offset: -0.1px;
  background-color: #FFFFFF;
  overflow: hidden;
}

.c3 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 16px;
  height: 16px;
  bottom: -4px;
  left: -4px;
  border-radius: 50%;
  outline: 2px solid #FFFFFF;
  outline-offset: -0.1px;
  background-color: #FFFFFF;
  overflow: hidden;
}

.c2 {
  width: 16px;
  height: 16px;
}

@media (max-width:960px) {
  .c0 {
    -webkit-align-items: flex-end;
    -webkit-box-align: flex-end;
    -ms-flex-align: flex-end;
    align-items: flex-end;
  }
}

@supports (overflow:clip) {
  .c1 {
    overflow: clip;
  }
}

@supports (overflow:clip) {
  .c3 {
    overflow: clip;
  }
}

<div
  class="c0"
  data-testid="StatusIconRoot"
  size="16"
>
  <div
    style="height: 16px; width: 16px; position: relative;"
  >
    <div
      style="height: 16px; width: 16px; overflow: visible; position: absolute;"
    >
      <svg
        viewBox="0 0 16 16"
      >
        <defs>
          <defs>
            <mask
              id="container-mask0x52270d8234b864dcAC9947f510CE9275A8a116Db16"
            >
              <rect
                fill="white"
                height="100%"
                width="100%"
                x="0"
                y="0"
              />
              <g
                transform="scale(0.4444444444444444) 
 translate(0, 0)"
              >
                <path
                  d="M18.1309 3.25957C9.91898 3.40293 3.14567 10.1762 3.00231 18.3882C2.85896 26.6001 9.39985 33.141 17.6118 32.9977C25.8238 32.8543 32.5971 26.081 32.7404 17.8691L33 3L18.1309 3.25957Z"
                  fill="black"
                />
              </g>
            </mask>
            <mask
              id="shape-mask0x52270d8234b864dcAC9947f510CE9275A8a116Db16"
            >
              <rect
                fill="white"
                height="100%"
                width="100%"
                x="0"
                y="0"
              />
              <g
                transform="scale(0.4444444444444444) 
 translate(10, 10)"
              >
                <path
                  clip-rule="evenodd"
                  d="M13.6569 13.6568C12.0059 10.0663 12.0059 5.93368 13.6569 2.34314C10.0663 3.99414 5.93368 3.99414 2.34315 2.34314C3.99414 5.93368 3.99414 10.0663 2.34315 13.6568C5.93368 12.0059 10.0663 12.0059 13.6569 13.6568ZM8 11C9.65685 11 11 9.65686 11 8.00001C11 6.34315 9.65685 5.00001 8 5.00001C6.34315 5.00001 5 6.34315 5 8.00001C5 9.65686 6.34315 11 8 11Z"
                  fill="black"
                  fill-rule="evenodd"
                />
              </g>
            </mask>
            <mask
              id="mask0x52270d8234b864dcAC9947f510CE9275A8a116Db16"
            >
              <g
                fill="white"
              >
                <g
                  mask="url(#shape-mask0x52270d8234b864dcAC9947f510CE9275A8a116Db16)"
                >
                  <g
                    transform="scale(0.4444444444444444)"
                  >
                    <path
                      d="M18.1309 3.25957C9.91898 3.40293 3.14567 10.1762 3.00231 18.3882C2.85896 26.6001 9.39985 33.141 17.6118 32.9977C25.8238 32.8543 32.5971 26.081 32.7404 17.8691L33 3L18.1309 3.25957Z"
                    />
                  </g>
                </g>
                <g
                  mask="url(#container-mask0x52270d8234b864dcAC9947f510CE9275A8a116Db16)"
                >
                  <g
                    transform="scale(0.4444444444444444)
        translate(10, 10)"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M13.6569 13.6568C12.0059 10.0663 12.0059 5.93368 13.6569 2.34314C10.0663 3.99414 5.93368 3.99414 2.34315 2.34314C3.99414 5.93368 3.99414 10.0663 2.34315 13.6568C5.93368 12.0059 10.0663 12.0059 13.6569 13.6568ZM8 11C9.65685 11 11 9.65686 11 8.00001C11 6.34315 9.65685 5.00001 8 5.00001C6.34315 5.00001 5 6.34315 5 8.00001C5 9.65686 6.34315 11 8 11Z"
                      fill-rule="evenodd"
                    />
                  </g>
                </g>
              </g>
            </mask>
          </defs>
          <lineargradient
            id="gradient0x52270d8234b864dcAC9947f510CE9275A8a116Db16"
          >
            <stop
              offset="0%"
              stop-color="#36DBFF"
            />
            <stop
              offset="100%"
              stop-color="#B8C3B7"
            />
          </lineargradient>
          <filter
            height="200%"
            id="blur0x52270d8234b864dcAC9947f510CE9275A8a116Db16"
            width="200%"
            x="-50%"
            y="-50%"
          >
            <fegaussianblur
              in="SourceGraphic"
              stdDeviation="5.333333333333333"
            />
          </filter>
        </defs>
        <g
          mask="url(#mask0x52270d8234b864dcAC9947f510CE9275A8a116Db16)"
        >
          <rect
            fill="url(#gradient0x52270d8234b864dcAC9947f510CE9275A8a116Db16)"
            height="100%"
            width="100%"
            x="0"
            y="0"
          />
          <rect
            fill="black"
            height="100%"
            opacity="0.08"
            width="100%"
            x="0"
            y="0"
          />
          <ellipse
            cx="8"
            cy="0"
            fill="#9D99F5"
            filter="url(#blur0x52270d8234b864dcAC9947f510CE9275A8a116Db16)"
            rx="8"
            ry="8"
          />
        </g>
      </svg>
    </div>
  </div>
  <div
    class="c1"
  >
    <img
      alt="Install MetaMask icon"
      class="c2"
      src="metamask-icon.svg"
    />
  </div>
  <div
    class="c3"
  >
    <img
      class="c2"
      src="socks.svg"
    />
  </div>
</div>
`;

exports[`StatusIcon with no account renders children in correct order 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-right: 8px;
}

.c0 > img,
.c0 span {
  height: 16px;
  width: 16px;
}

.c1 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 16px;
  height: 16px;
  bottom: -4px;
  right: -4px;
  border-radius: 50%;
  outline: 2px solid #FFFFFF;
  outline-offset: -0.1px;
  background-color: #FFFFFF;
  overflow: hidden;
}

.c3 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 16px;
  height: 16px;
  bottom: -4px;
  left: -4px;
  border-radius: 50%;
  outline: 2px solid #FFFFFF;
  outline-offset: -0.1px;
  background-color: #FFFFFF;
  overflow: hidden;
}

.c2 {
  width: 16px;
  height: 16px;
}

@media (max-width:960px) {
  .c0 {
    -webkit-align-items: flex-end;
    -webkit-box-align: flex-end;
    -ms-flex-align: flex-end;
    align-items: flex-end;
  }
}

@supports (overflow:clip) {
  .c1 {
    overflow: clip;
  }
}

@supports (overflow:clip) {
  .c3 {
    overflow: clip;
  }
}

<div
  class="c0"
  data-testid="StatusIconRoot"
  size="16"
>
  <div
    class="c1"
  >
    <img
      alt="Install MetaMask icon"
      class="c2"
      src="metamask-icon.svg"
    />
  </div>
  <div
    class="c3"
  >
    <img
      class="c2"
      src="socks.svg"
    />
  </div>
</div>
`;
